design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+
+We should also consider integrating metacontroller's functionality for storage backends that are compatible and can benefit from its patterns.
+
+### Source code and container images
+
+As more storage backends are integrated into Rook, it is preferable that all source code lives within the single `rook/rook` repository.
+This has a number of benefits such as easier sharing of build logic, developer iteration when updating shared code, and readability of the full source.
+Multiple container images can easily be built from the single source repository, similar to how `rook/rook` and `rook/toolbox` are currently built from the same repository.
+
+* `rook/rook` image:  defines all custom resource types, generated clients, general cluster discovery information (disks), and all custom controllers.  This image will be used for the Rook operator that hosts the backend specific in-proc controllers.
+* backend specific images: to avoid image bloat in the main `rook/rook` image, each backend will have their own image that contains all of their daemons and tools.  These images will be used for the various daemons/components of each backend, e.g. `rook/ceph`, `rook/minio`, etc.
+
+#### Layout
+
+The current `rook/rook` repository layout appears to be sufficiently factored to enable multiple storage backend support.
+Some additional directories will be added to support new API versions, new custom resource types, and new storage backends.
+A source code layout that includes these new additions is shown below, annotated with comments about the use of each important directory:
+
+```console
+- cmd # binaries with main entry points
+  - rook
+  - rookflex
+- pkg
+  - apis
+    - rook.io # rook.io API group, additional groups would be sibling dirs
+      - v1alpha1 # existing version of alpha Rook API
+      - v1alpha2 # new version of alpha Rook API that includes new types
+  - client
+    - clientset # generated strongly typed client code to access Rook API
+  - daemon # daemons for each storage backend
+    - ceph
+    - minio
+  - operator # all orchestration logic
+    - cluster # custom controllers for each storage backend
+      - ceph
+      - minio
+```
+
+### Custom Types
+
+Custom resources in Kubernetes use the following naming and versioning convention:
+
+* Group: A collection of several related types that can be enabled/disabled as a unit in the API (e.g., `rook.io`)
+* Version: The API version of the group (e.g., `v1alpha1`)
+* Kind: The specific type within the API group (e.g., `cluster`)
+
+Putting this together with an example, the `cluster` kind from the `rook.io` API group with a version of `v1alpha1` would be referred to in full as `cluster.rook.io/v1alpha1`.
+
+#### StorageScope
+
+The `storageScope.rook.io/v1alpha2` type defines the boundaries or "scope" of the resources that comprise the backing storage substrate for a cluster.
```

Comment:
Actually this may not be a top level CRD type.  It could just be a `StorageScopeSpec` type in golang only and not exposed as a CRD.  If exposed as a CRD, it would not be inlined in the user authored yaml, it would need to be a named reference to a separately defined and created `StorageScope` instance, right? Similar to PVCs:
```yaml
spec:
  scope:
    scopeName: myscope
```

Reply:
The complexity on "replicas" with storage clusters comes from the fact that there are two fundamental  strategies on how to protect storage from the loss of persistent drives: whole replicas and erasure coding. Some storage clusters will even seek to maintain a "must be protected from the loss of N drives" while changing between whole replicas and erasure coding as part of the lifecycle of any object.

There are also very subtle differences on erasure-coding specs depending on the exact erasure coding algorithm being used. I think that favors making it storage-cluster specific.

Reply:
I think I introduced some confusion by using the word "replicas", as there is a distinction to be made between number of storage daemon instances and the data redundancy/protection policy chosen.  I have changed the name of `replicas` to be `nodeCount` because what that field was trying to capture was at the cluster deployment level: how many nodes should I schedule an instance of the storage daemon to?

For data pools (after a storage cluster has been deployed), the user is more than welcome to specify a specific data protection policy, such as 3 "replicas" of each data object, or a K=2 M =1 erasure coding scheme.

Reply:
Since we discussing failure domain scope.. it would be nice to have a declarative way of assigning nodes to zones (or racks). For instance, NexentaEdge backend would benefit from the knowledge that backend point started in zone A and zone policy is actually enabled. As for erasure coding, failure domain schema would impose certain limitations as you rightly mentioned. For instance, if zoning is enabled, then in terms of to enable EC 4:2, one would need to have backend DS spanning across 6 zones.

Reply:
The `StorageScope` has the concept of a `Location` field.  This has been used so far to help build Ceph's crush map, but it is intended to be a generic construct for specifying topology information for resources included in the storage cluster (e.g., nodeA is in zone1 row2 rack3)